File: sequence.js
const midi = require('midi');
const util = require('./util');

class Sequence {
    constructor(name) {
        this.queue = [];
        this.output = new midi.output();
        util.selectPort(this.output, name);
    }

    pulse() {
        if(this.queue.length === 0) {
            this.next();
        }
        var message = this.queue.shift();
        if(typeof message !== "undefined") {
            this.output.sendMessage(message);
        }
    }

    next() {
        this.queue.push([0x99,0x3c,90]);
        this.queue.push(...Array(23));
    }

    start() {
        this.reset();
        this.next();
    }
    
    stop() {
        this.reset();
    }

    reset() {
        this.queue = [];
    }
}

module.exports = Sequence;

File: sequence.rb
class Sequence
  def initialize(name)
    @queue = []
    @midiout = RtMidi::Out.new
    port  = @midiout.port_names.index(name)
    if port.nil?
      puts midiout.port_names.inspect
      raise "no such device #{name}" 
    end
    @midiout.open_port(port)
  end

  def pulse
    steps if @queue.length == 0
    msg = @queue.shift
    @midiout.send_channel_message(*msg) unless msg.nil?
  end

  def steps
    @queue.push([0x99, 0x3c, 90])
    @queue.push(*([nil] * 23))
  end

  def reset
    @queue = []
  end
  
  def start
    reset
    steps
  end

  alias_method :stop,  :reset
end
